File: core/trading/jito_tips.py
"""
Dynamic Jito Tip Calculator.

Replaces static tip values with context-aware calculations based on:
    - Projected profit
    - Market volatility regime
    - Whether this is a sniper entry (time-sensitive)

Rules:
    - Never tip more than 20% of projected profit
    - Never use a static hardcoded tip value
    - Scale tips with volatility and urgency

Usage::

    from core.trading.jito_tips import calculate_jito_tip

    tip_sol = calculate_jito_tip(
        projected_profit_usd=50.0,
        market_volatility="medium",
        is_sniper_entry=False,
        sol_price_usd=150.0,
    )
"""

from __future__ import annotations

import logging
from enum import Enum
from typing import Optional

logger = logging.getLogger(__name__)


class VolatilityRegime(Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    EXTREME = "extreme"  # Graduation events, flash crashes


# Baseline tips by volatility regime (in SOL)
BASELINE_TIPS_SOL = {
    VolatilityRegime.LOW: 0.001,
    VolatilityRegime.MEDIUM: 0.004,
    VolatilityRegime.HIGH: 0.015,
    VolatilityRegime.EXTREME: 0.04,
}

# Maximum tip caps by regime (in SOL)
MAX_TIPS_SOL = {
    VolatilityRegime.LOW: 0.005,
    VolatilityRegime.MEDIUM: 0.01,
    VolatilityRegime.HIGH: 0.02,
    VolatilityRegime.EXTREME: 0.05,
}

# Profit-proportional caps by regime
PROFIT_TIP_FRACTION = {
    VolatilityRegime.LOW: 0.10,       # Tip up to 10% of profit
    VolatilityRegime.MEDIUM: 0.15,    # 15%
    VolatilityRegime.HIGH: 0.15,      # 15%
    VolatilityRegime.EXTREME: 0.20,   # 20%
}

# Hard cap: NEVER tip more than this fraction of projected profit
MAX_PROFIT_TIP_FRACTION = 0.20


def calculate_jito_tip(
    projected_profit_usd: float,
    market_volatility: str = "low",
    is_sniper_entry: bool = False,
    sol_price_usd: float = 150.0,
) -> float:
    """
    Calculate the optimal Jito tip in SOL.

    Args:
        projected_profit_usd: Expected profit from this trade in USD
        market_volatility: "low", "medium", "high", or "extreme"
        is_sniper_entry: True if this is a time-sensitive sniper trade
        sol_price_usd: Current SOL price for USD conversion

    Returns:
        Recommended tip in SOL
    """
    try:
        regime = VolatilityRegime(market_volatility.lower())
    except ValueError:
        regime = VolatilityRegime.LOW

    if sol_price_usd <= 0:
        sol_price_usd = 150.0

    # Start with baseline for the volatility regime
    tip = BASELINE_TIPS_SOL[regime]

    # Sniper entries need higher tips for priority landing
    if is_sniper_entry:
        tip = max(tip, BASELINE_TIPS_SOL.get(VolatilityRegime.HIGH, 0.015))
        if regime == VolatilityRegime.EXTREME:
            tip = BASELINE_TIPS_SOL[VolatilityRegime.EXTREME]

    # Scale with projected profit
    if projected_profit_usd > 0:
        profit_fraction = PROFIT_TIP_FRACTION[regime]
        profit_based_tip = (projected_profit_usd * profit_fraction) / sol_price_usd
        tip = max(tip, profit_based_tip)

    # Apply hard caps
    # 1. Never exceed regime-specific SOL cap
    max_sol = MAX_TIPS_SOL[regime]
    tip = min(tip, max_sol)

    # 2. Never tip more than 20% of projected profit
    if projected_profit_usd > 0:
        max_from_profit = (projected_profit_usd * MAX_PROFIT_TIP_FRACTION) / sol_price_usd
        tip = min(tip, max_from_profit)

    # 3. Minimum floor — always tip at least 0.0005 SOL to get included
    tip = max(tip, 0.0005)

    logger.debug(
        "Jito tip: %.6f SOL (regime=%s, sniper=%s, profit=$%.2f)",
        tip, regime.value, is_sniper_entry, projected_profit_usd,
    )
    return round(tip, 9)


def tip_to_lamports(tip_sol: float) -> int:
    """Convert SOL tip to lamports."""
    return int(tip_sol * 1_000_000_000)

File: jarvis-sniper/src/lib/jito-tips.ts
/**
 * Dynamic Jito Tip Calculator
 *
 * Calculates optimal Jito tips for Solana transactions based on:
 * - Projected profit from the trade
 * - Market volatility
 * - Whether it's a sniper entry (high priority)
 *
 * This replaces static hardcoded tips with dynamic calculation.
 *
 * @module jito-tips
 */

export type MarketVolatility = 'low' | 'medium' | 'high' | 'extreme';

export interface JitoTipConfig {
  /** Projected profit in USD */
  projectedProfitUsd: number;
  /** Current market volatility */
  volatility: MarketVolatility;
  /** Is this a sniper entry (launch sniping) */
  isSniperEntry: boolean;
  /** Current SOL price in USD */
  solPriceUsd: number;
  /** Minimum tip in SOL (default 0.001) */
  minTipSol?: number;
  /** Maximum tip as fraction of projected profit (default 0.20) */
  maxProfitFraction?: number;
}

// Base tip levels in SOL
const BASE_TIP_BY_VOLATILITY: Record<MarketVolatility, number> = {
  low: 0.001,      // 0.001 SOL
  medium: 0.003,   // 0.003 SOL
  high: 0.005,     // 0.005 SOL
  extreme: 0.01,   // 0.01 SOL
};

// Sniper entry multiplier
const SNIPER_MULTIPLIER = 2.0;

// Maximum tip as fraction of projected profit
const DEFAULT_MAX_PROFIT_FRACTION = 0.20;

/**
 * Calculate optimal Jito tip for a transaction
 *
 * Rules:
 * - Baseline (low volatility, non-sniper): 0.001 SOL
 * - Medium volatility: 0.003-0.005 SOL
 * - High volatility / sniper: min(0.02 SOL, projected_profit × 0.15 / SOL_price)
 * - Extreme volatility (graduation events): min(0.05 SOL, projected_profit × 0.20 / SOL_price)
 * - Never tip more than 20% of projected profit
 *
 * @param config - Jito tip configuration
 * @returns Optimal tip in SOL
 */
export function calculateJitoTip(config: JitoTipConfig): number {
  const {
    projectedProfitUsd,
    volatility,
    isSniperEntry,
    solPriceUsd,
    minTipSol = 0.001,
    maxProfitFraction = DEFAULT_MAX_PROFIT_FRACTION,
  } = config;

  // 1. Start with base tip based on volatility
  let tipSol = BASE_TIP_BY_VOLATILITY[volatility];

  // 2. Apply sniper multiplier for high-priority entries
  if (isSniperEntry) {
    tipSol *= SNIPER_MULTIPLIER;
  }

  // 3. Calculate maximum tip based on projected profit
  const maxTipByProfit = (projectedProfitUsd / solPriceUsd) * maxProfitFraction;

  // 4. Apply volatility scaling for high/extreme volatility
  if (volatility === 'high') {
    tipSol = Math.min(tipSol, 0.02); // Cap at 0.02 SOL for high volatility
  } else if (volatility === 'extreme') {
    tipSol = Math.min(tipSol, 0.05); // Cap at 0.05 SOL for extreme volatility
  }

  // 5. Ensure we don't exceed profit-based maximum
  tipSol = Math.min(tipSol, maxTipByProfit);

  // 6. Ensure minimum tip
  tipSol = Math.max(tipSol, minTipSol);

  return Math.round(tipSol * 1000000) / 1000000; // Round to 6 decimal places
}

/**
 * Calculate tip for a specific use case
 */
export function calculateTipForUseCase(
  useCase: 'normal' | 'sniper' | 'graduation' | 'emergency',
  projectedProfitUsd: number,
  solPriceUsd: number = 150
): number {
  const volatilityMap: Record<string, MarketVolatility> = {
    normal: 'low',
    sniper: 'high',
    graduation: 'extreme',
    emergency: 'extreme',
  };

  return calculateJitoTip({
    projectedProfitUsd,
    volatility: volatilityMap[useCase],
    isSniperEntry: useCase === 'sniper' || useCase === 'graduation',
    solPriceUsd,
  });
}

/**
 * Estimate tip in USD for display
 */
export function tipToUsd(tipSol: number, solPriceUsd: number): number {
  return tipSol * solPriceUsd;
}

/**
 * Get tip recommendation for UI display
 */
export function getTipRecommendation(
  tradeSizeSol: number,
  solPriceUsd: number,
  volatility: MarketVolatility = 'medium'
): {
  tipSol: number;
  tipUsd: number;
  description: string;
} {
  const tradeSizeUsd = tradeSizeSol * solPriceUsd;

  // Estimate 5% profit for tip calculation
  const estimatedProfit = tradeSizeUsd * 0.05;

  const tipSol = calculateJitoTip({
    projectedProfitUsd: estimatedProfit,
    volatility,
    isSniperEntry: false,
    solPriceUsd,
  });

  return {
    tipSol,
    tipUsd: tipToUsd(tipSol, solPriceUsd),
    description: `${volatility} volatility tip recommendation`,
  };
}
